### DISCUSSION TITLE
Uploading Big File using FastAPI

### DISCUSSION THREAD

**Discussion by jponnetCytomine** [fastapi/fastapi#9828] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter, File, UploadFile
from starlette.requests import Request

router = APIRouter()

@router.post('/upload/')
async def chunk_upload(
    request: Request
):
    upload_path = PENDING_PATH
    pending_path = Path(upload_path)

    async with aiofiles.open(pending_path, 'wb') as f:
        async for chunk in request.stream():
            await f.write(chunk)

@router.post('/upload/UploadFile')
async def UploadFile_upload(
    file: UploadFile = File(...)
):
    upload_path = PENDING_PATH
    file_content = await file.read()
    pending_path = Path(upload_path)

    with open(pending_path, "wb") as f:
        f.write(file_content)
```

### Description

I am using postman to do some upload tests on either /upload/UploadFile or /upload

The purpose of the two functions is to upload a file. However, one upload a file using the fastAPI "UploadFile" and the other uses directly the request. 

In my application, we upload very large images. I tried using UploadFile, but it is rather slow compared to other alternatives (using module in NGINX for file upload) . After an intensive investigation, I noticed that the code lines where it takes time is in the routing.py function `body = await request.form()` because it writes the whole file which is in the request body. 

I found an alternative that is using `request.stream()`, it works much better although I encountered a problem. When I write the chunks into the file on my disk, the first lines of the first chunks looks like 
`----------------------------125063113267490430295207\r\nContent-Disposition: form-data; name="files[].file"; filename="test.png"\r\nContent-Type: image/png\r\n\r\n\`
and if i write those lines in the file, then I can not open the image (it does not recognize this file as a png image) 
For now I have arranged my code such that I does not write these kind of body headers (start writing after the empty `\r\n\r\n`) but this is not a robust solution. 

Do you have an alternative? Am I doing something wrong? Can you explain me more in details why using UploadFile is so slow? 

Thank you very much for your help! 

### Operating System

Linux

### FastAPI Version

0.99.1

### Python Version

3.8

**Comment by mikeedjones** [comment, marked as answer]
I believe that the header row you're seeing is a consequence of the _upload_ and not on fastAPI or starlette's side. https://www.rfc-editor.org/rfc/rfc7578#section-4.2

You can avoid adding by using the data field it (assuming you're using requests to upload files)

```
response = requests.put(url, headers=headers, data=open(fpath,'rb').read())
```
